openthread: Enable nrf_security support for precompiled OT libs (#2974)
nrfconnect/sdk-nrf: doc/nrf/ug_thread_configuring.rst

  You can set any value ranging from ``0`` to ``65535``.

Default configuration reference

>     The default configuration for all :ref:`openthread_samples` is defined in :file:`nrf/samples/common/overlay-ot-defaults.conf`.

Author: ```suggestion
    The default configuration for all :ref:`openthread_samples` is defined in :file:`nrf/samples/openthread/common/overlay-ot-defaults.conf`.
```

Selecting these sets is not related to :ref:`thread_ug_device_type`.

.. note::

>     Notice that while building OpenThread from sources an additional user configuration can overwrite selected sets of features.

Reviewer: Can we be more strict here?
I propose something like that:
```suggestion
    While building OpenThread from sources the user configuration overwrites selected predefined Master, FTD or MTD feature set.
```

*****************************

|NCS| provides predefined sets of features and optional functionalities from the OpenThread stack.

> These predefined sets are based on :ref:`nrfxlib:ot_libs` available in nrfxlib.

Author: I would rather say that :ref:`nrfxlib:ot_libs` provide functionalities listed in these predefined sets.

> You can use these sets for configuration purposes or for building libraries with complete Thread specification support when you :ref:`configure OpenThread to use pre-built libraries <ug_thread_configuring_basic_building>` with the :option:`CONFIG_OPENTHREAD_LIBRARY_1_1` Kconfig option.

Author: ```suggestion
You can use these sets for configuration purposes when you build your application using OpenThread sources or for building application with complete Thread specification support when you :ref:`configure OpenThread to use pre-built libraries <ug_thread_configuring_basic_building>` with the :option:`CONFIG_OPENTHREAD_LIBRARY_1_1` Kconfig option.
```

=======================================

You can update nrfxlib's :ref:`nrfxlib:ot_libs` when using any Thread sample if you configure the sample to build OpenThread stack from source with :option:`CONFIG_OPENTHREAD_SOURCES`.
Use this functionality for example for :ref:`certification <ug_thread_cert>` of your configuration of OpenThread libraries.

.. note::
    The libraries destination directory can differ.
    When you selected :option:`CONFIG_OPENTHREAD_USER_CUSTOM_LIBRARY`, the location depends the chosen :ref:`nrf_security backend <nrfxlib:nrf_security_readme>`, either :option:`CONFIG_CC310_BACKEND` or :option:`CONFIG_OBERON_BACKEND`.

Updating libraries without debug symbols
----------------------------------------

You can install the release version of the latest nrfxlib libraries without the debug symbols to save a lot of memory space.
This is handled with the :option:`CONFIG_OPENTHREAD_BUILD_OUTPUT_STRIPPED` Kconfig option.
This option is disabled by default.

Run the following command to update the nrfxlib libraries:

.. parsed-literal::
   :class: highlight

   west build -b nrf52840dk_nrf52840 -t install_openthread_libraries -- -DOPENTHREAD_BUILD_OUTPUT_STRIPPED=y

>    This command builds two versions of the libraries, with and without debug symbols, and installs only the version without debug symbols.
> 

Reviewer 2: ```suggestion

This command builds two versions of the libraries, with and without debug symbols, and installs only the version without debug symbols.
```

      - ✔
      -
      -
      -
    * - DHCP6_CLIENT
      - ✔
      - ✔
      - ✔
      -
    * - ECDSA
      - ✔
      - ✔
      - ✔
      -
    * - IP6_FRAGM
      - ✔
      - ✔
      - ✔
      -
    * - JAM_DETECTION
      - ✔
      - ✔
      - ✔
      -
    * - JOINER
      - ✔
      - ✔
      - ✔
      -
    * - LINK_RAW
      - ✔
      -
      -
      -
    * - MAC_FILTER
      - ✔
      - ✔
      - ✔
      -
    * - MTD_NETDIAG
      - ✔
      -
      -
      -
    * - SERVICE
      - ✔
      - ✔
      -
      -
    * - SLAAC
      - ✔
      - ✔
      - ✔
      -
    * - SNTP_CLIENT
      - ✔
      - ✔
      - ✔
      -
    * - UDP_FORWARD
      - ✔
      - ✔
      -
      -

.. _thread_ug_feature_updating_libs:

Updating pre-built OpenThread libraries
=======================================

You can update nrfxlib's :ref:`nrfxlib:ot_libs` when using any Thread sample if you configure the sample to build OpenThread stack from source with :option:`CONFIG_OPENTHREAD_SOURCES`.
Use this functionality for example for :ref:`certification <ug_thread_cert>` of your configuration of OpenThread libraries.

.. note::
    The libraries destination directory can differ.

>     When you selected :option:`CONFIG_OPENTHREAD_USER_CUSTOM_LIBRARY`, the location depends the chosen :ref:`nrf_security backend <nrfxlib:nrf_security_readme>`, either :option:`CONFIG_CC310_BACKEND` or :option:`CONFIG_OBERON_BACKEND`.

Reviewer 3: CONFIG_CC3XX_BACKEND

  west build -t install_openthread_libraries

.. note::

>     The libraries destination directory differ based on chosen option :option:`CONFIG_OPENTHREAD_NORDIC_LIBRARY` and cryptography backends: :option:`CONFIG_CC310_BACKEND` or :option:`CONFIG_OBERON_BACKEND`.

Reviewer: Awesome!

      - ✔
      -

Replacing OpenThread libraries
==============================

Each OpenThread sample provides a possibility to replace nrfxlib's :ref:`nrfxlib:ot_libs` when configured to build OpenThread stack from source with :option:`CONFIG_OPENTHREAD_SOURCES`.

> After the sample is built, run the following command in the sample directory to overwrite the existing nrfxlib libraries with the new ones:

Author: Actually you don't need to run the command always in the sample directory. For instance you can just run:
`west build -b nrf52840dk_nrf52840 -p always nrf/samples/openthread/cli/` from root directory and there you can just run `west build -t install_openthread_libraries` without moving to sample directory.

      - ✔
      -

.. _thread_ug_feature_updating_libs:

Updating OpenThread libraries
=============================

Each OpenThread sample provides a possibility to update nrfxlib's :ref:`nrfxlib:ot_libs` when the sample is configured to build OpenThread stack from source with :option:`CONFIG_OPENTHREAD_SOURCES`.
You can use this functionality for certification with your own custom libraries.

After a sample is built, run the following command to overwrite the update nrfxlib libraries to your latest version:

.. parsed-literal::
   :class: highlight

   west build -t install_openthread_libraries

.. note::
    The libraries destination directory can differ.

>     When you selected :option:`CONFIG_OPENTHREAD_USER_CUSTOM_LIBRARY`, the location depends the chosen nRF Security backend, either :option:`CONFIG_CC310_BACKEND` or :option:`CONFIG_OBERON_BACKEND`.

Reviewer 4: Something to consider: Maybe we should not mention backends here and mention it "depends on device hardware crypto capabilities" as it is more complex than just CC and OBERON i.e. on nRF53 we'll have CC312, and on 52833 currently we use mixed config oberon + vanilla due to lack of SHA224 support in oberon(to be added soon).
